### ISSUE TITLE
outerIIFEBody does not handle all IIFE cases

### ISSUE THREAD

**Issue by TheCycoONE** [eslint/eslint#6613]
**What version of ESLint are you using?**
3.0.1

**What parser (default, Babel-ESLint, etc.) are you using?**
espree

**Please show your full configuration:**
```js
module.exports = {
    "env": {
        "browser": true,
        "amd": true
    },
    "extends": "eslint:recommended",
    "rules": {
        "indent": [
            "error",
            "tab",
            { "outerIIFEBody": 0 }
        ],
        "linebreak-style": [
            "error",
            "windows"
        ],
        "semi": [
            "error",
            "always"
        ],
        "no-unused-vars": [
            "error",
            { "vars": "local", "args": "after-used" }
        ],
        "no-empty": [
            "error",
            { "allowEmptyCatch": true }
        ]
    },
    "globals": {
        "dojo": false,
        "ActiveXObject": false
    }
};
```

**What did you do? Please include the actual source code causing the issue.**
```js
window.ERMS.FloatingDialog = (function(evts) {
"use strict";
var my;
my.foo = function() {
	return evts;
};

return my;
})(window.ERMS.Events);
```

**What did you expect to happen?**
No errors

**What actually happened? Please include the actual, raw output from ESLint.**
```
  2:1  error  Expected indentation of 1 tab character but found 0  indent
  3:1  error  Expected indentation of 1 tab character but found 0  indent
  4:1  error  Expected indentation of 1 tab character but found 0  indent
  8:1  error  Expected indentation of 1 tab character but found 0  indent
```

I believe this case would be covered if the false negatives mentioned by @wojdyr in https://github.com/eslint/eslint/issues/6585#issuecomment-230183366 were handled.

Pasting those examples here for convenience:
```js
(function () {
var i;  // ok
})();

~function () {
var i;  // not
}();

!(function () {
var i;  // not
})();

!function () {
var i;  // not
}();

;(function () {
var i;  // ok
})();

var MyClass = (function () {
var i;  // not
})();
```

**Comment by platinumazure** [comment]
Labeled as enhancement request (under the assumption that with @nzakas' latest bug fix, the rule is now working "as designed", whatever that means). If this should be re-labeled as a bug, please feel free to correct as needed.

**Comment by nzakas** [comment]
@TheCycoONE we never intended to support assignment as in your example, and I'm not entirely sure it makes sense to do so because the IIFE isn't really "outer" at that point. I can see an argument for supporting unary operators before the IIFE, but I'm not sure about assignment.

**Comment by TheCycoONE** [comment]
Alright. We're using module pattern so the top of the file contains the assignment to the namespace. We use large (vim default 8 space tab) indentation and avoid indenting the entire file. Similar to Google's goog.scope we don't indent for namespaces. Perhaps a new *namespace* indent option could catch both of those cases?

wojdyr mentioned the Google case as his original concern in https://github.com/eslint/eslint/issues/6259 where it appears that the outerIIFEBody solution came from.

**Comment by wojdyr** [comment]
I added the examples with unary operators just for completeness, but I'm also interested only in the case with assignment.

I thought that *outer* refers to namespace (i.e. outer = not in another function) not to AST.

**Comment by nzakas** [comment]
The original purpose for this option was to support the jQuery style guide:
https://contribute.jquery.org/style-guide/js/#full-file-closures

And, once I looked back at that, it seems that they do support assignments so we should as well.

I'll champion this change.

@TheCycoONE are you willing to implement this once accepted?

**Comment by ilyavolodin** [comment]
:+1:

**Comment by markelog** [comment]
Also, created issue on jquery config repo - https://github.com/jquery/eslint-config-jquery/issues/4

**Comment by mikesherov** [comment]
:+1:

**Comment by mysticatea** [comment]
:+1:

**Comment by TheCycoONE** [comment]
I'm not familiar with the code base, but I'll take a look on Monday and let you know if I can figure it out.
